flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/BinaryToBinaryCastRule.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.functions.casting;
+
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeFamily;
+import org.apache.flink.table.types.logical.utils.LogicalTypeChecks;
+
+import java.util.Arrays;
+
+import static org.apache.flink.table.planner.functions.casting.CastRuleUtils.staticCall;
+
+/** {@link LogicalTypeFamily#BINARY_STRING} to {@link LogicalTypeFamily#BINARY_STRING} cast rule. */
+public class BinaryToBinaryCastRule
+        extends AbstractExpressionCodeGeneratorCastRule<byte[], byte[]> {
+
+    static final BinaryToBinaryCastRule INSTANCE = new BinaryToBinaryCastRule();
+
+    private BinaryToBinaryCastRule() {
+        super(
+                CastRulePredicate.builder()
+                        .input(LogicalTypeFamily.BINARY_STRING)
+                        .target(LogicalTypeFamily.BINARY_STRING)
+                        .build());
+    }
+
+    @Override
+    public String generateExpression(
+            CodeGeneratorCastRule.Context context,
+            String inputTerm,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        int inputLength = LogicalTypeChecks.getLength(inputLogicalType);
+        int targetLength = LogicalTypeChecks.getLength(targetLogicalType);
+
+        if (targetLength >= inputLength) {
+            return inputTerm;
+        } else {
+            return staticCall(Arrays.class, "copyOfRange", inputTerm, 0, targetLength);
+        }
```

Comment:
Unfortunately you can't assume that the input length is the same as the one defined in the type. In our codegen domain we can have a `byte[]` of length 10 with type defined as `VARBINARY(20)` or even `BINARY(20)`. If you look carefully, even the `RawToBinaryCastRule` you introduced behaves the same when casting to `BINARY(20)` a raw which array has just size 10.

Hence you need the check at runtime to avoid an unnecessary copy, for example in this case:

* input `VARBINARY(30)` with value `byte[].length == 10`
* target `VARBINARY(20)`

The first branch of the if is fine IMO, because it's reasonable to assume that if the input length is minor or equal to the target length, then it's fine to just return the same term. The second branch on the other hand should do a length check with the ternary operator and skip the copy in case the runtime length is already less or equal to the target length.

Reply:
Thx @slinkydeveloper! I totally missed that.

Reply:
Good point! I will update here to check during runtime